### PR TITLE
[TRAVIS] Use 'build stages' on travis and restructure build steps

### PR DESCRIPTION
**Pull request by andreas-bok-sociomantic**
Migrate the .travis.yml config file to use
build stages.

Sepearates compilation/testing and deployment into
build stages.

Add user-defined cache directory to avoid building
the deb packages in the deployment stage

### CODE REVIEW

**Comment by andreas-bok-sociomantic** [comment]
@leandro-lucarella-sociomantic identical to #266 but on the `v5.7.x` branch. Would rather have the build stage changes in both v5.7.x and v6.1.x. If it LGTY then feel free to merge

File: .travis.yml [hunk 1]
```diff
@@ -1,33 +1,26 @@
 language: cpp
 
 ## Enable cache for dependencies and directories
-cache: ccache
+cache:
+  directories:
+    - build/travis-cache
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: I would just use `$CACHE_DIR` here to keep it consistent.

File: travis_configure.sh [hunk 2]
```diff
@@ -22,6 +22,30 @@ print_error_msg ()
...
+    fi
+
+    case $1 in
+        jfrog)
+            echo "Installing $1"
+            curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > /tmp/jfrog ;
+            chmod a+x /tmp/jfrog ;
+            mkdir -p $HOME/bin ;
+            cp /tmp/jfrog $HOME/bin/jfrog ;
+            #export PATH=$HOME/bin:$PATH
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh]: I would remove this commented-out line.

File: travis_configure.sh [hunk 3]
```diff
@@ -22,6 +22,30 @@ print_error_msg ()
     return 0
 }
 
+install_dependency()
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh]: Why not just `install_jfrog` if there is nothing to reuse here?

File: travis_configure.sh [hunk 4]
```diff
@@ -38,12 +62,8 @@ before_install()
...
 
         # jfrog dependency
-        if [[ "$MAKE_TARGET" =~ "deb" ]]; then
-            if [[ -n "$TRAVIS_TAG" && $TRAVIS_REPO_SLUG == "sociomantic-tsunami/libdrizzle-redux" ]]; then
-                curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > /tmp/jfrog ;
-                chmod a+x /tmp/jfrog ;
-                sudo cp /tmp/jfrog /usr/local/bin/jfrog ;
-            fi
+        if [[ -n "$TRAVIS_TAG" ]]; then
+            install_dependency jfrog
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh]: Thy is this duplicated? You are also calling this in the `.travis.yml` file.

File: travis_configure.sh [hunk 5]
```diff
@@ -137,6 +157,9 @@ elif [[ $1 == "before_script" ]]; then
     before_script
 elif [[ $1 == "run_tests" ]]; then
     run_tests
+elif [[ $1 == "install_dependency" ]];  then
+    shift
+    install_dependency $1
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh]: And why the forwarding? I guess this goes beyond this PR, but it seems to add a lot of complexity.

Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: The reason is that running the shared `before_install` script in the deployment step would build the docker image which isn't needed for just uploading the `deb` packages to `bintray`
Furthermore the shared  `before_script` and `script` are short-circuited by just exiting in at the end of `deploy.before_install`

Hmm, could move all stuff in `travis_configure.sh::before_install()`  to `travis_configure.sh::before_script()` and have `travis_configure.sh::before_install()` only install dependencies for the travis trusty build environment.

File: travis_configure.sh (hunk 4 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: Cf. https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/275/files#r162381250

File: travis_configure.sh (hunk 3 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: Cf. https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/275/files#r162381250

File: travis_configure.sh (hunk 4 above)
Comment (leandro-lucarella-sociomantic) [reply travis_configure.sh]: This link is not sending me to anywhere useful (just takes me to this same thread?). Also, what does `Cf.` mean?

File: travis_configure.sh (hunk 5 above)
Comment (leandro-lucarella-sociomantic) [reply travis_configure.sh]: OK, I'm lost, too much complexity for me to follow, but it's not really part of this PR so nevermind.

File: travis_configure.sh (hunk 4 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: > Cf. https://github.com/sociomantic-tsunami/libdrizzle-redux/pull/275/files#r162381250

Sorry linking to the longer comment below.

Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: cf. == confer

**Comment by andreas-bok-sociomantic** [comment]
Split PR into separate commits

File: travis_configure.sh [hunk 6]
```diff
@@ -38,12 +38,10 @@ before_install()
...
-        if [[ "$MAKE_TARGET" =~ "deb" ]]; then
-            if [[ -n "$TRAVIS_TAG" && $TRAVIS_REPO_SLUG == "sociomantic-tsunami/libdrizzle-redux" ]]; then
-                curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > /tmp/jfrog ;
-                chmod a+x /tmp/jfrog ;
-                sudo cp /tmp/jfrog /usr/local/bin/jfrog ;
-            fi
+        if [[ -n "$TRAVIS_TAG" ]]; then
+            curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > /tmp/jfrog ;
+            chmod a+x /tmp/jfrog ;
+            sudo cp /tmp/jfrog /usr/local/bin/jfrog ;
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh]: Are these 2 changes really related, couldn't you change the conditions without touching this file?

File: travis_configure.sh [hunk 7]
```diff
@@ -24,19 +24,12 @@ print_error_msg ()
...
 #   - sed, (libtool), openssl, mysql
 # Returns 0
 before_install()
 {
     if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
-        # build images
-        docker-compose -f ./docker-compose.yml \
-            -f ./docker/docker-compose.$DIST_NAME.$DIST_VERSION.yml \
-            build
-
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh:39]: Why is this moved?

File: travis_configure.sh [hunk 8]
```diff
@@ -32,9 +32,11 @@ before_install()
     if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
         # jfrog dependency
         if [[ -n "$TRAVIS_TAG" ]]; then
-            curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > /tmp/jfrog ;
-            chmod a+x /tmp/jfrog ;
-            sudo cp /tmp/jfrog /usr/local/bin/jfrog ;
+            if [[ ! -d $HOME/bin || ! -e $HOME/bin/jfrog ]]; then
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh]: This is a bit redundant, `[[ -e $HOME/bin/jfrog ]]` should be enough, right? Also, why do you even bother to check? Who do you expect to install `jfrog` in the home directory for you?

File: .travis.yml [hunk 9]
```diff
@@ -91,6 +94,8 @@ jobs:
     - <<: *linux-build
       env:
         COMPILER_VERSION=4.9 MAKE_TARGET=$MAKE_TARGET:deb
+      before_cache:
+        - if [[ -n $TRAVIS_TAG ]]; then cp pkg/deb/*.deb $CACHE_DIR/; fi
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: This probably needs quotes: `[[ -n "$TRAVIS_TAG" ]]` otherwise if the tag is empty it could give a syntax error (at least on some shell implementations).

File: .travis.yml [hunk 10]
```diff
@@ -103,10 +108,13 @@ jobs:
...
       stage: deploy
       if: tag IS present AND tag =~ ^v[0-9]+\.[0-9]+\.[0-9]+
       env:
         COMPILER_VERSION=4.9 MAKE_TARGET=$MAKE_TARGET:deb
-      after_success:
-        - beaver bintray upload -D $DIST_VERSION -N pkg/deb/*.deb
+      before_script: skip
+      script:
+        - beaver bintray upload -D $DIST_VERSION -N $CACHE_DIR/*.deb
+        - exit 0
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: Why the `exit 0`?

File: travis_configure.sh (hunk 7 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh:39]: The intention is that `before_install` only installs dependencies on the **travis** build environment. 
Thus it allows us to do `before_script: skip` if it isn't needed to build the docker images. As is the case in the deploy stage where the deb packages are uploaded.

File: travis_configure.sh (hunk 6 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: No, need to remove the **TRAVIS_REPO_SLUG** since `jfrog` wouldn't be available in a tagged build from a fork.

File: travis_configure.sh (hunk 8 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: > This is a bit redundant, 

Sure, was thinking it would print the error like doing `ls` on a non-existing nested directory

File: .travis.yml (hunk 10 above)
Comment (andreas-bok-sociomantic) [reply .travis.yml]: Wanting to finish the build as soon as possible. But see the point that it isn't required since build steps after `script`, `after_success` etc, have been defined. Will remove

File: travis_configure.sh (hunk 7 above)
Comment (leandro-lucarella-sociomantic) [reply travis_configure.sh:39]: Maybe explain that in the commit message? :)
(and maybe even a comment in the travis file, so people seeing it know why it's organized like it's organized)

File: .travis.yml (hunk 10 above)
Comment (leandro-lucarella-sociomantic) [reply .travis.yml]: exit 0 should really be a NOP. If the previous command failed, it won't reach it, if it didn't, it will end up successfully anyway :)

File: travis_configure.sh (hunk 8 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: >  Also, why do you even bother to check? Who do you expect to install jfrog in the home directory for you?

So you mean removing any checks and installing `jfrog` as default?

**Comment by andreas-bok-sociomantic** [comment]
Addressed comments by luca 

Previous head was 26a8b82, new **HEAD** at 9387778.

<details>
<summary>Diff between heads are:</summary>

```diff

 .travis.yml         |  3 +--
 travis_configure.sh | 16 ++++++++++------
 2 files changed, 11 insertions(+), 8 deletions(-)

diff --git a/.travis.yml b/.travis.yml
index b395df4..7abf23a 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -95,7 +95,7 @@ jobs:
       env:
         COMPILER_VERSION=4.9 MAKE_TARGET=$MAKE_TARGET:deb
       before_cache:
-        - if [[ -n $TRAVIS_TAG ]]; then cp pkg/deb/*.deb $CACHE_DIR/; fi
+        - if [[ -n "$TRAVIS_TAG" ]]; then cp pkg/deb/*.deb $CACHE_DIR/; fi
     - <<: *linux-build
       env:
         CXX=clang++ CC=clang COMPILER_VERSION=3.9
@@ -117,4 +117,3 @@ jobs:
       before_script: skip
       script:
         - beaver bintray upload -D $DIST_VERSION -N $CACHE_DIR/*.deb
-        - exit 0
diff --git a/travis_configure.sh b/travis_configure.sh
index e80dbe4..63bf208 100755
--- a/travis_configure.sh
+++ b/travis_configure.sh
@@ -24,6 +24,9 @@ print_error_msg ()
 
 # Script which is run before the installation script is called
 #
+# Only dependencies required in the native travis build environment should be
+# installed in this step.
+#
 # For osx based builds homebrew is used to install required packages:
 #   - sed, (libtool), openssl, mysql
 # Returns 0
@@ -31,12 +34,10 @@ before_install()
 {
     if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
         # jfrog dependency
-        if [[ -n "$TRAVIS_TAG" ]]; then
-            if [[ ! -d $HOME/bin || ! -e $HOME/bin/jfrog ]]; then
-                mkdir -p $HOME/bin ;
-                curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > $HOME/bin/jfrog ;
-                chmod a+x $HOME/bin/jfrog ;
-            fi
+        if [[ -n "$TRAVIS_TAG" && ! -e $HOME/bin/jfrog ]]; then
+            mkdir -p $HOME/bin ;
+            curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > $HOME/bin/jfrog ;
+            chmod a+x $HOME/bin/jfrog ;
         fi
     elif [[ "$TRAVIS_OS_NAME" == "osx" ]]; then
         brew update
@@ -76,6 +77,9 @@ enable_mysqlbinlog()
 #
 # For linux based builds docker-compose is used to set up the build environment
 #
+# No dependencies required in the native travis build environment should be
+# installed in this step. Use before_install instead
+#
 # Returns 0 or 1 if called with an invalid build configuration
 before_script()
 {

```
</details>

**Comment by andreas-bok-sociomantic** [comment]
Ready to go, @leandro-lucarella-sociomantic ?

File: .travis.yml [hunk 11]
```diff
@@ -85,14 +45,67 @@ before_script:
...
+    - &linux-build
+      services:
+        - docker
+    # Global config for builds on osx
+    - &osx-build
+      os: osx
+      compiler: clang
+    # Global config for deployment stage
+    - &deploy
+      os: linux
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: Sorry if this was there long ago, but I just see it.

These templates look really funny, it seems what they include looks pretty arbitrary and also the OSX and deploy template are only used once. I would  only move stuff to a template when you really need to reuse it, otherwise it just makes following this file more complicated.

You can leave it for a future PR though.

File: travis_configure.sh (hunk 8 above)
Comment (leandro-lucarella-sociomantic) [reply travis_configure.sh]: Yeah, but it's not important.

**Review by leandro-lucarella-sociomantic — APPROVED** [review]

**Comment by leandro-lucarella-sociomantic** [comment]
Some more minor comments but they can be addressed in another PR, no need to block this.

File: .travis.yml (hunk 11 above)
Comment (andreas-bok-sociomantic) [reply .travis.yml]: > These templates look really funny, it seems what they include looks pretty arbitrary 

You the `services.docker`? The reason for that is that docker isn't needed in the deploy stage

Comment (andreas-bok-sociomantic) [reply .travis.yml]: >  also the OSX and deploy template are only used once

OSX: I know, but basically just to prepare it for building with different versions of XCode
Deploy: There is another PR for building coverage (`test-stage`) and uploading (`deploy-stage`) it to `coveralls.io`.

**Comment by andreas-bok-sociomantic** [comment]
> Some more minor comments but they can be addressed in another PR, no need to block this.

Great
